### PR TITLE
Fix for join mode.

### PR DESCRIPTION
**Pull request by dbonadiman**
I need to introduce a bit the problem first.
The `join` mode for the merge layer is the layer designed to pass the inputs of a graph node to the layer without any transformation. That was the pull request with the relative discussion.

https://github.com/fchollet/keras/pull/690

The problem with the previous implementation of the code was that It always produce the exception `raise ValueError("merge_mode='join' only works with named inputs")`
since there are no named output tensors in keras. 

```python
 for i in range(len(self.layers)):
     X = self.layers[i].get_output(train)
    if X.name is None:
       raise ValueError("merge_mode='join' only works with named inputs")
    else:
       inputs[X.name] = X
 return inputs
```

The first and simple solution that came to my mind was obviously to name any output tensors in the code. But i think that for the moment is dispendious and useless.

So i decided to add a parameter `name` to the layer itself that is set with the set_name function.

Tell me what you ( @fchollet  ) and @EderSantana (the original creator of the join mode) think and in particular if any further modification is needed.

### CODE REVIEW

**Comment by EderSantana** [comment]
Right now, join only works if you have are inside a layer Graph. This is will make sure all your variables get properly named. I believe it is a plan to make all models named.

**Comment by dbonadiman** [comment]
@EderSantana  Yeah, i got this point. But right now the Graph doesn't name properly all the layers. It uses the name internally for storage purposes and it applies the set name function to the layer that only names the parameters. However the output is still unnamed so your implementation.
, besides is more correct, fails most of the times. In particular where you merge a symmetric Graph, outputs are often called  `mean`, `sum`as the basic theano functions and the produced Ordereddict often contains only one value.

**Comment by EderSantana** [comment]
That naming convention you're proposing is actually interesting and easier to control. Also it makes the model independent of Theano naming, which should be useful for future backends. 

One thing, did you test if `join` now works with Sequential models? Would you have time for writing a little unittest for this PR?

**Comment by dbonadiman** [comment]
@EderSantana  At the moment unfortunately not. I've only tested only on the Graph model.

And i suppose that it doesn't work in sequential models since they are not named. 
But it was working before? I don't understand how. They had not naming system either.

**Comment by EderSantana** [comment]
nice! That is what I expected.

Also from my experience with other PRs, some docs and or tests used to be required... specially for interesting features that enhance model capabilities somehow. It can be a simple unittest like the following https://github.com/fchollet/keras/blob/master/tests/auto/test_graph_model.py#L22-L43 using a model like that you just wrote. Make it compile and assert it returns an output to `predict`. Doesn't need extensive testing.

**Comment by dbonadiman** [comment]
@EderSantana missed these kind of tests i should only check if the loss decreases?
I think i will check them with the new LambdaMerge.

However there is a point missing to fix, the output shape.
I used the `join` mode paired with some layers created by me where i had to pass a new parameter `input_shapes=[model.nodes['layer1'].output_shape, model.nodes['layer2'].output_shape]` in this way since the join mode output only a None `output_shape`.

**Comment by EderSantana** [comment]
@dbonadiman you don't have to even see the loss decreasing. it is enough to be able to compile the model and call the `predict` method. Tests are also simple forms of tutorials sometimes. Write the test to answer the question "how can one use merge_mode=`join` with my brand new feature". Just write a simple example and yes, you can use the LambdaLayer or design an appropriate layer inside the text, I'd rather go with Lambdas for simplicity. I think that anybody not lazy enough always go to tests to learn about the library. I personally believe tests > docs sometimes because they give working examples.  

This is one thing you should put on a comment in your test: "there is no general way to do shape inference when using merge_mode=`join`, you will have to define the input_dim of the next layer yourself". I believe this is not a problem since any user experienced enough to be able to use `join` and write his own Keras layers should be able to calculate an `input_dim` or whatever is necessary to build the layer after `join`, don't worry too much to complain with the shape inference if you can't. `join`, and LambdaLayers are about flexibility and new ideas.

Finally, the PR would be more than perfect if you copy paste your code to the `Merge` layer examples topic in the docs: https://github.com/fchollet/keras/blob/master/docs/sources/layers/core.md

**Comment by EderSantana** [comment]
@fchollet I believe @dbonadiman had a neat and simple solution for naming conventions that is independent of the way we do it in Theano. It makes `merge_mode=join` just work even for Sequential, which is a problem @farizrahman4u and I had before. Please, give it a read to see if conforms with the plans for the new backends and naming conventions you had in mind. I'm doing the coding review, I just needed your opinion to avoid giving suggestions that won't get ultimately accepted.

@dbonadiman let us just wait for François real quick so you don't work without being assured this is good for Keras.

File: keras/layers/core.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ def __init__(self, **kwargs):
             self._trainable = kwargs['trainable']
         if not hasattr(self, 'params'):
             self.params = []
+        self.name = None
```
Comment (fchollet) [inline keras/layers/core.py]: Then, why not allow a `name` keyword argument that would set `self.name`?

**Comment by fchollet** [comment]
> @fchollet I believe @dbonadiman had a neat and simple solution for naming conventions that is independent of the way we do it in Theano. It makes merge_mode=join just work even for Sequential, which is a problem @farizrahman4u and I had before. Please, give it a read to see if conforms with the plans for the new backends and naming conventions you had in mind. I'm doing the coding review, I just needed your opinion to avoid giving suggestions that won't get ultimately accepted.

That's cool. Go ahead, if there are conflicts in the future we will fix them in the future.

**Comment by dbonadiman** [comment]
There is even another modification in mind @fchollet @EderSantana , that requires a bit of refactor of the code but not of the apis. And may be included in this pull request.

A decorator like that (probably better):
```python
def check_output(func):
    def func_wrapper(self, *args, **kwargs):
        X = func(self, *args, **kwargs)
        if list(X.shape)[1:] == list(self.output_shape)[1:]
             X.name = '_'.join([self.name, 'output')
        else:
             raise Exception('the output shape of the layer is not the one expected')
```
That needs to be applied on the get_output method:

```python
@check_output
def get_output(self):
    ...
```

I think that is an elegant solution to solve the output naming problem and checks on the output_shape that are currently missing from keras.

I don't think it will works directly on all the layers for instance the merge join outputs Ordered dicts that can't be named so i need to verify.

**Comment by dbonadiman** [comment]
Fixed some things here and there to allows naming directly in the constructor of the layer and added documentation

**Comment by EderSantana** [comment]
You lost me now, why do you need that decorator? The other layers do not need to support input coming from `join`.

**Comment by fchollet** [comment]
Yes, that decorator is pretty obscure imho.

**Comment by dbonadiman** [comment]
I will do a separate pull request if needed

**Comment by dbonadiman** [comment]
The decorator idea was for doing standard operation on the output of a layer:
   -  Check if the actual output shape match the output shape encoded in the output_shape function
   -  Name the tensor in output.

It's kind of a related problem to the one i encountered in implementing this solution for the `join`, the output of the layer was not named. In that sense was necessary to modify any `get_output` function to set a name on the tensor from one side or avoid all the checks on names as i did.
Pairing the decorator with the `self.name` parameter i created allows us to set the name of the tensor at runtime.

But as previously said i will avoid adding the decorators for now, maybe i can decore some code in a separate PR ;)

**Comment by dbonadiman** [comment]
@fchollet @EderSantana the decorator i presented here there was wrong here is a naming wrapper that works.
https://github.com/dbonadiman/keras/blob/patch-6.1/keras/utils/decorators_util.py

I added it to all the layers and run the current tests (it works). However i don't know how much overhead is added on the get_output function.

what do you think? is it something needed?

**Comment by dbonadiman** [comment]
I've added the BiLinearLayer to make real examples with `join`:

```
W1 dot x1 + W2 dot x2 + b
```

**Comment by EderSantana** [comment]
The bilinear layer is not the right way to do this, is it? If you `concat` x1 and x2 and do a single matrix multiplication, it will run much faster than split it into two smaller multiplications. I don't see it as a good use case of `join`.

I think I understand how the decorator works now, but do you mind writing a little snippet of a problem case that this decorator solves? Is that an alternative to your previous solution?

**Comment by fchollet** [comment]
Closing outdated PRs following the release of Keras 1.0.
